### PR TITLE
Upgrade ec2_remote_facts to use boto3

### PR DESCRIPTION
**Pull request by wimnat**
##### SUMMARY
Update ec2_remote_facts to boto3

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2_remote_facts

##### ANSIBLE VERSION
2.4

##### ADDITIONAL INFORMATION
ec2_remote_facts misses some important information when describing instances and also misses returns doc.

This PR:

 - Update to use boto3
 - Paginate results so if there are hundreds of instances returned, this will be handled
 - Add documented return values

### CODE REVIEW

**Comment by willthames** [comment]
Can we rename it while we're at it to `ec2_instance_facts` and deprecate `ec2_remote_facts`? At some point we can rename `ec2_facts` to `ec2_instance_metadata_facts` or something.

File: lib/ansible/modules/cloud/amazon/ec2_remote_facts.py [hunk 1]
```diff
@@ -42,148 +49,496 @@
...
 
-    for instance in all_instances:
-        instance_dict_array.append(get_instance_info(instance))
+    # Get instances from reservations
+    instances = []
+    for reservation in reservations['Reservations']:
+        instances = instances + reservation['Instances']
+
+    # Turn the boto3 result in to ansible_friendly_snaked_names
+    snaked_instances = []
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_remote_facts.py]:
```
snaked_instances = [camel_dict_to_snake_dict(instance) for instance in instances]
```

**Review by willthames — CHANGES_REQUESTED** [review]
Given that this is almost certainly backward compatible with respect to return data, I'm going to insist that the module be renamed (which is a great opportunity anyway)

File: lib/ansible/modules/cloud/amazon/ec2_remote_facts.py [hunk 2]
```diff
@@ -42,148 +49,496 @@
...
+    instance_ids = module.params.get("instance_ids")
+    filters = ansible_dict_to_boto3_filter_list(module.params.get("filters"))
 
     try:
-        all_instances = connection.get_only_instances(filters=filters)
-    except BotoServerError as e:
+        reservations_paginator = connection.get_paginator('describe_instances')
+        reservations = reservations_paginator.paginate(InstanceIds=instance_ids, Filters=filters).build_full_result()
+    except ClientError as e:
         module.fail_json(msg=e.message)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_remote_facts.py]: Might as well move to full boto3 exception handling standards

**Comment by wimnat** [comment]
@willthames I completely agree regarding the rename.  I've implemented that and also added in the updated exception handling.

**Comment by willthames** [comment]
lgtm - I'll test this in the next day or so, thanks for the fixes @wimnat

**Comment by wimnat** [comment]
@willthames did you get around to testing this one?

**Comment by wimnat** [comment]
shipit

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
Rebased, looks good to ship once tests pass

**Pull request merged**